syndeo-template/src/main/java/com/google/cloud/syndeo/common/ProviderUtil.java

```diff
@@ -0,0 +1,120 @@
+    ServiceLoader<SchemaTransformProvider> provider =
+        ServiceLoader.load(SchemaTransformProvider.class);
+    List<SchemaTransformProvider> list =
+        StreamSupport.stream(provider.spliterator(), false).collect(Collectors.toList());
+    list.addAll(SchemaIOTransformProviderWrapper.getAll());
+
+    Map<String, SchemaTransformProvider> map = new HashMap<>();
+    for (SchemaTransformProvider p : list) {
+      map.put(p.identifier(), p);
+    }
+
+    return map;
+  }
+
+  /** Return a list of all providers. */
+  public static Collection<SchemaTransformProvider> getProviders() {
+    return PROVIDERS.values();
+  }
+
+  /** Get a specific provider. */
+  public static SchemaTransformProvider getProvider(String name) {
+    return PROVIDERS.getOrDefault(name, null);
+  }
+
+  /** A class for storing an io / configuration pair. */
+  public static class TransformSpec {
+    public String inputId;
+    public Row configuration = null;
+    public SchemaTransformProvider provider = null;
+
+    public TransformSpec(String inputId, List<Object> configurationAsList) {
+      this.inputId = inputId;
+      provider = getProvider(inputId);
+      if (provider == null) throw new RuntimeException(inputId + " not found ");
+      configuration =
+          Row.withSchema(provider.configurationSchema()).addValues(configurationAsList).build();
+    }
+
+    public TransformSpec(ConfiguredSchemaTransform schemaTransform) {
+      inputId = schemaTransform.getTransformUrn();
+      provider = getProvider(inputId);
+      if (provider == null) throw new RuntimeException(inputId + " not found ");
+      // TODO: Test handling of rows with missing fields.
+      configuration =
+          SchemaUtil.addNullsToMatchSchema(
+              (Row)
+                  SchemaTranslation.rowFromProto(
+                      ProtoTranslation.fromSyndeoProtos(schemaTransform.getConfigurationValues()),
+                      FieldType.row(provider.configurationSchema())),
+              provider.configurationSchema());
+    }
+
+    public ConfiguredSchemaTransform toProto() {
+      ConfiguredSchemaTransform.Builder inst = ConfiguredSchemaTransform.newBuilder();
+      inst.setConfigurationValues(
+          ProtoTranslation.toSyndeoProtos(SchemaTranslation.rowToProto(configuration)));
+      inst.setConfigurationOptions(
+          ProtoTranslation.toSyndeoProtos(
+              SchemaTranslation.schemaToProto(configuration.getSchema(), true)));
+      inst.setTransformUrn(inputId);
+      return inst.build();
+    }
+  }
+
+  /** Applies the given configs. */
+  public static PCollectionRowTuple applyConfigs(
+      Collection<TransformSpec> specs, PCollectionRowTuple tuple) {
+    for (TransformSpec spec : specs) {
+      SchemaTransform transform = spec.provider.from(spec.configuration);
+      // We know we only deal with transforms with either 0 or 1 input so we know how to connect the
```

Comment:
More a note: I don't know if this will be the case long-term.

Reply:
agreed. I'll leave it as-is but we may change this over time.

Reply:
Ack.